src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorPullDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Diagnostics;
+
+internal class RazorPullDiagnosticsEndpoint : IRazorPullDiagnosticsEndpoint
+{
+    public bool MutatesSolutionState => false;
+
+    public RegistrationExtensionResult GetRegistration(VSInternalClientCapabilities clientCapabilities)
+    {
+        return new RegistrationExtensionResult("_vs_supportsDiagnosticRequests", options: clientCapabilities.SupportsDiagnosticRequests);
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalDocumentDiagnosticsParams request)
+    {
+        if (request.TextDocument is null)
+        {
+            throw new NotImplementedException("Why would document be null?");
+        }
+
+        return request.TextDocument;
+    }
+
+    public async Task<IEnumerable<VSInternalDiagnosticReport>?> HandleRequestAsync(VSInternalDocumentDiagnosticsParams request, RazorRequestContext context, CancellationToken cancellationToken)
+    {
+        var result = await GetPullDiagnosticsAsync(request, context, cancellationToken);
```

Comment:
Need to check single server, right? And no-op if its turned off?

Why not just inherit from our base delegating single server endpoint thingy? There is less benefit from spreading out the single server work, so everyone has a go, without also getting feedback back into the system about shortcomings of the design.

Reply:
> Need to check single server, right?

I suppose I had imagined this was taken care of by capabilities. I'm moving to the base branch in my response anyway.

Reply:
At the moment none of the single server endpoints do it that way, but some of them, this included, could. Probably a better method too.